### DISCUSSION TITLE
Templating

### DISCUSSION THREAD

**Discussion by medonja** [NUMde/compass-numapp-frontend#141] [Q&A] [answered]
Hi there,

I saw that recently a "transform into a template" has been conducted for the frontend. Can you tell  a bit more about the template, what is it all about? It doesn't seem to be explained anywhere in the docs? :)

I have tried following the new [initial setup instructions](https://github.com/NUMde/compass-numapp-frontend/tree/main/docs/setup), but it seems that the instruction for replacing the X.X.X with a recent template version is not working, since the provided URL is dead https://github.com/mbastian93/release-test/releases

Thanks in advance for the support! :)

**Comment by mbastian93** [comment]
Hey

First of all the link to the releases was wrong; it should be fixed now :)

Since this project is now a template, a new project based upon _this project_ is not created by cloning this repository but instead by invoking the command described in the documentation. You can then configure and modify your version of the app according to your needs and likings.

Should you want to contribute to the template, you can fork it as before and then submit your pull request.

**Comment by medonja** [comment]
Hey, thanks for the update and given explanations, I appreciate it! :)

I have now tried to setup a new project according to the updated docs, but running the android app raised the following error. I have setup also a google-services.json file in the android/app directory. This is weird, as I haven't experienced any such issue with the non-template version of the frontend. Could you help me out with this one? Any support is appreciated!

![image](https://user-images.githubusercontent.com/9157689/169814619-fb86bb46-233b-40f1-9e0f-b782f32d6235.png)

**Comment by mbastian93** [comment, marked as answer]
That's a known issue; this file is not copied during the installation.

For now, please download [this file](https://github.com/NUMde/compass-numapp-frontend/blob/main/compass-numapp-template/android/app/google-services.json) manually and place it in the android/app folder

Edit:
When you want to use push notifications, you have to set it up according to the [documentation](https://github.com/NUMde/compass-numapp-frontend/tree/main/docs/setup#enabling-push-notifications). The file provided above acts just as placeholder so that the app can be built.

**Comment by medonja** [comment]
I have actually tried that already, and I also tried it with the placeholder google-services.json, and in both cases I get the aforementioned error message. Any ideas what else might have gone wrong?

**Reply by mbastian93** [reply]
The id in the google-services.json must match your app id.
in your case replace 'de.num.compass.android' with 'de.num.ukbumfragetool.android' in google-services.json

**Reply by medonja** [reply]
Ah you are absolutely right. I actually did that too, but had a typo which I did not notice. Alright, thanks a lot!

**Comment by medonja** [comment]
Maybe one more question: what is the best practice now to catch up with your updates to NUM Compass. Let's say we start developing our app from the 0.1.0 template, and then after a while we want to be up to date with your latest changes, bugfixes and improvements. Previously, we would pull the changes from github. What should we do now to achieve this?

**Reply by mbastian93** [reply]
> Maybe one more question: what is the best practice now to catch up with your updates to NUM Compass. Let's say we start developing our app from the 0.1.0 template, and then after a while we want to be up to date with your latest changes, bugfixes and improvements. Previously, we would pull the changes from github. What should we do now to achieve this?

Admittedly, there is no straightforward 'easy' way to achieve this. Even with the previous way some conflicts are not avoidable the more a custom version diverges from the 'original'. There are two ways you can incorporate changes made to the template to your own project:

1. manually copy changes; when there aren't many changes between the template and your project, you can manually copy the changes. This way, you can leave out changes, that don't affect your project.
2.  - create a new project based upon the newest version of the template (Alternatively, just download the source code from the newest release)
     - create a new branch on your original project and replace all* files with the content from the newly generated project
     - merge this branch into your main branch and resolve all resulting conflicts

---
\* Except config files etc
